vllm/model_executor/model_loader/loader.py

```diff
@@ -347,6 +347,124 @@ def load_model(self, *, model_config: ModelConfig,
                                              vision_language_config)
 
 
+class ShardedStateLoader(BaseModelLoader):
+    """
+    Model loader that directly loads each worker's model state dict, which
+    enables a fast load path for large tensor-parallel models where each worker
+    only needs to read its own shard rather than the entire checkpoint. See
+    `examples/save_sharded_states.py` for creating a sharded checkpoint.
+    """
+
+    DEFAULT_PATTERN = "model-rank-{rank}-part-{part}.safetensors"
+
+    def __init__(self, load_config: LoadConfig):
+        super().__init__(load_config)
+        extra_config = ({} if load_config.model_loader_extra_config is None
+                        else load_config.model_loader_extra_config.copy())
+        self.pattern = extra_config.pop("pattern", self.DEFAULT_PATTERN)
+        if extra_config:
+            raise ValueError(f"Unexpected extra config keys for load format "
+                             f"{load_config.load_format}: "
+                             f"{load_config.model_loader_extra_config.keys()}")
+
+    @staticmethod
+    def _filter_subtensors(tensors: Dict[str, torch.Tensor]):
+        """
+        Filter out all tensors that share the same memory or a subset of the
+        memory of another tensor.
+        """
+        from safetensors.torch import storage_ptr, storage_size
+        result = {}
+        for key1, tensor1 in tensors.items():
```

Comment:
this looks suspect from a performance perspective (O(n^2)) but I guess it should be fine...

Reply:
this should be fine -- as a model at most just has 100 tensors..

Reply:
Updated this part as there were some other issues. Now we only do the n^2 loop over tensors that share the same underlying storage, which should be much fewer than before.